nixnet/types.py

```diff
@@ -495,6 +496,130 @@ def __repr__(self):
+        "eventslot",
+        "eventid",
+        "payload"]
+
+    _FRAME_ID_MASK = 0x0000003F
+
+    def __init__(self, identifier, type=constants.FrameType.LIN_DATA, payload=b""):
+        # type: (int, constants.FrameType, bytes) -> None
+        self.identifier = identifier
+        self.echo = False  # Used only for Read
+        self.type = type
+        self.timestamp = 0  # Used only for Read
+        self.eventslot = False  # Used only for Read
+        self.eventid = 0  # Used only for Read
+        self.payload = payload
+
+    @classmethod
+    def from_raw(cls, frame):
+        # type: (RawFrame) -> LinFrame
+        """Convert from RawFrame.
+
+        >>> raw = RawFrame(5, 2, constants.FrameType.LIN_DATA, _cconsts.NX_FRAME_FLAGS_TRANSMIT_ECHO, 0, b'\x01')
+        >>> LinFrame.from_raw(raw)
+        LinFrame(identifier=2, echo=True, timestamp=0x5, len(payload)=1)
+        """  # NOQA: E501
+        identifier = frame.identifier & cls._FRAME_ID_MASK
+        lin_frame = LinFrame(identifier, constants.FrameType(frame.type), frame.payload)
+        lin_frame.timestamp = frame.timestamp
+        lin_frame.echo = bool(frame.flags & _cconsts.NX_FRAME_FLAGS_TRANSMIT_ECHO)
+        lin_frame.eventslot = bool(frame.flags & _cconsts.NX_FRAME_FLAGS_LIN_EVENT_SLOT)
+        if lin_frame.eventslot:
+            lin_frame.eventid = frame.info
+        else:
+            lin_frame.eventid = 0
+
+        return lin_frame
+
+    def to_raw(self):
+        # type: () -> RawFrame
+        """Convert to RawFrame.
+
+        >>> LinFrame(2, constants.FrameType.LIN_DATA).to_raw()
+        RawFrame(timestamp=0x0, identifier=0x2, type=FrameType.LIN_DATA)
+        >>> l = LinFrame(2, constants.FrameType.LIN_DATA)
+        >>> l.echo = True
+        >>> l.to_raw()
+        RawFrame(timestamp=0x0, identifier=0x2, type=FrameType.LIN_DATA, flags=0x80)
+        """
+        if self.identifier != (self.identifier & self._FRAME_ID_MASK):
+            _errors.check_for_error(_cconsts.NX_ERR_UNDEFINED_FRAME_ID)
+        flags = 0
+        info = 0
+        if self.echo:
+            flags |= _cconsts.NX_FRAME_FLAGS_TRANSMIT_ECHO
+        if self.eventslot:
+            flags |= _cconsts.NX_FRAME_FLAGS_LIN_EVENT_SLOT
+            info |= self.eventid
+        return RawFrame(self.timestamp, self.identifier, self.type, flags, info, self.payload)
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            other_frame = typing.cast(LinFrame, other)
+            return all((
+                self.identifier == other_frame.identifier,
+                self.echo == other_frame.echo,
+                self.type == other_frame.type,
+                self.timestamp == other_frame.timestamp,
+                self.eventslot == other_frame.eventslot,
+                self.eventid == other_frame.eventid,
+                self.payload == other_frame.payload))
+        else:
+            return NotImplemented
+
+    def __repr__(self):
```

Comment:
A couple parts of your repr are missing coverage
https://coveralls.io/builds/12752500/source?filename=nixnet%2Ftypes.py#L592